zerver/webhooks/newrelic/view.py

```diff
@@ -1,169 +1,278 @@
+        "total_incidents": str(payload["totalIncidents"].tame(check_int)),
+        "state": payload.get("state", "UPDATED").tame(check_string_in(INCIDENT_STATES)),
+        "status": determine_status(payload),
+        "priority": priority,
+        "priority_symbol": priority_symbol,
+    }
 
-TOPIC_TEMPLATE = """{policy_name} ({incident_id})""".strip()
+    return message_context
 
-# Once old is EOLed, delete old and keep new
-OLD_EVENT_TYPES = ["closed", "acknowledged", "open"]
-NEW_EVENT_TYPES = ["created", "activated", "acknowledged", "closed"]
-ALL_EVENT_TYPES = list(set(OLD_EVENT_TYPES).union(set(NEW_EVENT_TYPES)))
 
+def determine_status(payload: WildValue) -> str:
+    """
+    This function infers what type of action was taken on the incident based on the
+    timestamps provided in the payload.
 
-@webhook_view("NewRelic", all_event_types=ALL_EVENT_TYPES)
+    Returns the action taken as a string.
+    """
+    updated_at = TIME_STAMP_STR.get("UPDATED")
+    acknowledged_at = TIME_STAMP_STR.get("ACKNOWLEDGED")
+    created_at = TIME_STAMP_STR.get("CREATED")
+    closed_at = TIME_STAMP_STR.get("CLOSED")
+
+    if updated_at:
+        if acknowledged_at and updated_at == acknowledged_at:
+            return "ACKNOWLEDGED"
+        elif created_at and updated_at == created_at:
+            return "ACTIVATED"
+        elif closed_at and updated_at == closed_at:
+            return "CLOSED"
+        else:
+            return "UPDATED"
+    incident_title = payload.get("title", "No title.").tame(check_string)
+    raise UnsupportedWebhookEventTypeError(incident_title)
+
+
+def generate_additional_info_message(payload: WildValue) -> Optional[str]:
+    """
+    Parses additional fields from the recommended base payload
+    into a presentable block of strings.
+    Detail: https://zulip.com/integrations/doc/newrelic#recommended-base-payload-template
+
+    Returns a block of formatted strings.
+    """
+    additional_info_block = ""
+    owner = payload.get("owner").tame(check_none_or(check_string))
+    acknowledge_by_row = f"|:silhouette:  Acknowledged By **{owner}**|\n" if owner else ""
+
+    additional_info_block += acknowledge_by_row
+
+    return additional_info_block
+
+
+def generate_custom_field_message(payload: WildValue) -> Optional[str]:
+    """
+    Parses the custom fields in the "zulipCustomFields" dict in
+    the payload by turning the camel case key into words and
+    formatting it together with the value into a presentable
+    block of strings.
+    Detail:https://zulip.com/integrations/doc/newrelic#displaying-custom-fields
+
+    Returns a block of formatted strings.
+    """
+    body_custom_field_detail: str = ""
+    zulip_custom_fields = payload.get(ZULIP_CUSTOM_FIELDS_KEY, {})
+
+    for key, value in zulip_custom_fields.items():
+        camel_case_pattern = r"^[a-z]+(?=[A-Z])|[0-9]|[A-Z](?:[a-z]+|[A-Z]*(?=[A-Z]|$))"
+        tmp_name = re.findall(camel_case_pattern, key)
```

Comment:
Is there a reason that we're doing all this reformatting of the key and requiring camel case format? Why can't we just accept any string for the key and display it as such in the message?

Reply:
Yeah, I agree, that makes sense. we should probably just write somewhere in the doc that the key for these custom payloads will be used in the notification as is. So, something like this:
```
{
   ...
     "owner": "N/A",
    "zulipCustomFields": {
      "Your Custom Fields": "somedata123",
      "Custom Status 1": true,
      "Custom List 1": ["SSD", 2000, false, null, 13.33],
      "Custom Field 1": null,
      "Custom Int 2": 9000
    }
  }
```
Initially, I matched it to how Newrelic would format its payload, but there's no reason to do so for this particular case.